src/censored.jl

```diff
@@ -0,0 +1,409 @@
+continuous components.
+
+```julia
+censored(d0, l, missing)   # d0 left-censored to the interval [l, Inf)
+censored(d0, missing, u)   # d0 right-censored to the interval (-Inf, u]
+censored(d0, l, u)         # d0 interval-censored to the interval [l, u]
+```
+
+The function falls back to constructing a [`Censored`](@ref) wrapper.
+
+# Implementation
+
+To implement a specialized censored form for distributions of type `D`, one or more of the
+following methods should be implemented:
+- `censored(d::D, l::T, u::T) where {T <: Real}`
+- `censored(d::D, ::Missing, u::Real)`
+- `censored(d::D, l::Real, ::Missing)`
+"""
+censored
+function censored(d::UnivariateDistribution, l::T, u::T) where {T<:Real}
+    return Censored(d, l, u)
+end
+function censored(d::UnivariateDistribution, ::Missing, u::Real)
+    return Censored(d, missing, u)
+end
+function censored(d::UnivariateDistribution, l::Real, ::Missing)
+    return Censored(d, l, missing)
+end
+censored(d::UnivariateDistribution, l::Real, u::Real) = censored(d, promote(l, u)...)
+censored(d::UnivariateDistribution, ::Missing, ::Missing) = d
+
+"""
+    Censored
+
+Generic wrapper for a [`censored`](@ref) distribution.
+"""
+struct Censored{
+    D<:UnivariateDistribution,
+    S<:ValueSupport,
+    T<:Real,
+    TL<:Union{T,Missing},
+    TU<:Union{T,Missing},
+} <: UnivariateDistribution{S}
+    uncensored::D      # the original distribution (uncensored)
+    lower::TL      # lower bound
+    upper::TU      # upper bound
+    function Censored(d::UnivariateDistribution, l::T, u::T) where {T<:Real}
+        _is_non_empty_interval(l, u) ||
+            error("the lower bound must be less than or equal to the upper bound")
+        new{typeof(d), value_support(typeof(d)), T, T, T}(d, l, u)
+    end
+    function Censored(d::UnivariateDistribution, l::Missing, u::T) where {T <: Real}
+        new{typeof(d), value_support(typeof(d)), T, Missing, T}(d, l, u)
+    end
+    function Censored(d::UnivariateDistribution, l::T, u::Missing) where {T <: Real}
+        new{typeof(d), value_support(typeof(d)), T, T, Missing}(d, l, u)
+    end
+end
+
+function censored(d::Censored, l::T, u::T) where {T<:Real}
+    return censored(
+        d.uncensored,
+        d.lower === missing ? l : max(l, d.lower),
+        d.upper === missing ? u : min(u, d.upper),
+    )
+end
+function censored(d::Censored, ::Missing, u::Real)
+    return censored(d.uncensored, d.lower, d.upper === missing ? u : min(u, d.upper))
+end
+function censored(d::Censored, l::Real, ::Missing)
+    return censored(d.uncensored, d.lower === missing ? l : max(l, d.lower), d.upper)
+end
+
+function params(d::Censored)
+    d0params = params(d.uncensored)
+    return if d.lower === missing
+        (d0params..., d.upper)
+    elseif d.upper === missing
+        (d0params..., d.lower)
+    else
+        (d0params..., d.lower, d.upper)
+    end
+end
+
+partype(d::Censored) = partype(d.uncensored)
```

Comment:
I'd argue that `d.lower` and `d.upper` are parameters of the censored distribution, hence probably this should be
```suggestion
partype(d::Censored{<:UnivariateDistribution,<:ValueSupport,T}) where {T} = promote_type(partype(d.uncensored), T)
```
or maybe even just
```suggestion
partype(d::Censored{<:UnivariateDistribution,<:ValueSupport,T}) where {T} = T
```
Not sure if we should take into account the parameter type of nested distributions.

Reply:
I think if we're returning them in `params`, then we should.

Reply:
Seems reasonable.